contracts/governance/governance.cpp

```diff
@@ -0,0 +1,642 @@
+   system::detail::max_address_size,        // proposal.set_system_contract.contract_id
+   system::detail::max_signatures_length,   // proposal.signatures length
+   system::detail::max_signature_size >;    // proposal.signatures
+
+bool proposal_updates_governance( const koinos::system::transaction& proposal )
+{
+   for ( uint64_t i = 0; i < proposal.get_operations().get_length(); i++ )
+   {
+      const auto& op = proposal.get_operations().get_const( i );
+      switch( op.get_which_op() )
+      {
+         case koinos::system::operation::FieldNumber::UPLOAD_CONTRACT:
+         {
+            const auto& upload_op = op.get_upload_contract();
+
+            std::string contract_id( reinterpret_cast< const char* >( upload_op.get_contract_id().get_const() ), upload_op.get_contract_id().get_length() );
+
+            // We're uploading the current governance contract
+            if ( contract_id == system::get_contract_id() )
+               return true;
+         }
+         break;
+
+         case koinos::system::operation::FieldNumber::CALL_CONTRACT:
+         break;
+
+         case koinos::system::operation::FieldNumber::SET_SYSTEM_CALL:
+         {
+            const auto& set_system_call_op = op.get_set_system_call();
+
+            std::vector< koinos::chain::system_call_id > governance_syscalls;
+            governance_syscalls.push_back( koinos::chain::system_call_id::pre_block_callback );
+            governance_syscalls.push_back( koinos::chain::system_call_id::require_system_authority );
+            governance_syscalls.push_back( koinos::chain::system_call_id::apply_set_system_call_operation );
+            governance_syscalls.push_back( koinos::chain::system_call_id::apply_set_system_contract_operation );
+
+            for ( const auto& syscall : governance_syscalls )
+               if ( set_system_call_op.get_call_id().get() == std::underlying_type_t< koinos::chain::system_call_id >( syscall ) )
+                  return true;
+         }
+         break;
+
+         case koinos::system::operation::FieldNumber::SET_SYSTEM_CONTRACT:
+         {
+            const auto& set_system_contract_op = op.get_set_system_contract();
+
+            std::string contract_id( reinterpret_cast< const char* >( set_system_contract_op.get_contract_id().get_const() ), set_system_contract_op.get_contract_id().get_length() );
+
+            // We're setting the system status of the governance contract
+            if ( contract_id == system::get_contract_id() )
+               return true;
+         }
+         break;
+
+         default:
+         break;
+      }
+   }
+
+   return false;
+}
+
+std::vector< proposal_record > retrieve_proposals(
+   std::function< bool( const proposal_record& ) > predicate = []( const proposal_record& ) { return true; },
+   uint64_t limit = 0,
+   std::string start_proposal = std::string{} )
+{
+   std::vector< proposal_record > proposals;
+
+   // No limit
```

Comment:
Is no limit intentional? If so, why have it at all? If not, we need to set a hard limit.

Reply:
It is intentional. It's really based on the particular nodes `read_compute_limit`. I think it makes sense to leave this up to the developer and node operator.